pkg/collector/prometheus.go

```diff
@@ -0,0 +1,583 @@
+		prometheus.BuildFQName(namespace, "pod", "uncore_joules_total"),
+		"Aggregated RAPL value in uncore in joules",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+	podDramJoulesTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "dram_joules_total"),
+		"Aggregated RAPL value in dram in joules",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+	podPackageJoulesTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "package_joules_total"),
+		"Aggregated RAPL value in package (socket) in joules",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+	podOtherComponentsJoulesTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "other_host_components_joules_total"),
+		"Aggregated value in other host components (platform - package - dram) in joules",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+	podGPUJoulesTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "gpu_joules_total"),
+		"Aggregated GPU value in joules",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+
+	// Hardware Counters (counter)
+	podCPUCyclesTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "cpu_cycles_total"),
+		"Aggregated CPU cycle value",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+	podCPUInstrTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "cpu_instructions_total"),
+		"Aggregated CPU instruction value",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+	podCacheMissTotal := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "cache_miss_total"),
+		"Aggregated cache miss value",
+		[]string{"pod_name", "pod_namespace", "command"}, nil,
+	)
+
+	// Additional metrics (gauge)
+	podCPUTime := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "cpu_cpu_time_us"),
+		"Current CPU time per CPU",
+		[]string{"pod_name", "pod_namespace", "cpu"}, nil,
+	)
+
+	// Old deprecated metrics
+	podEnergyStat := prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "pod", "energy_stat"),
+		"Several labeled pod metrics (deprecated)",
+		podEnergyStatLabels, nil,
+	)
+
+	c.podDesc = &PodDesc{
+		podCoreJoulesTotal:            podCoreJoulesTotal,
+		podUncoreJoulesTotal:          podUncoreJoulesTotal,
+		podDramJoulesTotal:            podDramJoulesTotal,
+		podPackageJoulesTotal:         podPackageJoulesTotal,
+		podOtherComponentsJoulesTotal: podOtherComponentsJoulesTotal,
+		podGPUJoulesTotal:             podGPUJoulesTotal,
+		podCPUCyclesTotal:             podCPUCyclesTotal,
+		podCPUInstrTotal:              podCPUInstrTotal,
+		podCacheMissTotal:             podCacheMissTotal,
+		podCPUTime:                    podCPUTime,
+		podEnergyStat:                 podEnergyStat, // deprecated
+	}
+}
+
+// Collect implements the prometheus.Collector interface
+func (c *PrometheusCollector) Collect(ch chan<- prometheus.Metric) {
+	lock.Lock()
```

Comment:
do we still need this lock?

it 1st time of request will be blocked by wg until it finished.
so here comes with the 2nd request.
will the 2nd one waiting until 1st return? or it needs a lock for waiting 1st completed?

Reply:
The `lock` here is to synchronize with the metric collection
Reporting metrics in the middle of the `reader` function will create wrong data....
The wait is to make sure that all the gorotines reported the metrics when a Collect is triggered

Reply:
may I know where can I know more about how `reader ` and `Collect` works together?
I want to ensure multiple thread in safe and effective.

Reply:
I will use in this explanation the structure names as I refactored in PR

The `containersMetrics` contains all the metrics that the `reader` collects/calculates, e.g. energy consumption and resource utilization from the bpf

The `reader` is a gorotine that updates the `containersMetrics` every `SamplePeriodSec`, which is 3s

The `Collect` function implements the prometheus interface, which reports metrics on an endpoint in `\metrics`.
The prometheus app is typically configured with a default scrape interval and will get metrics from an endpoint according to that interval. But this interval can also be set in `servicemonitor`, in kepler we set this to be 3s, so it will match the `SamplePeriodSec`.....

So when prometheus tries to get data from kepler, the `Collect` function will be called and will access the `containersMetrics` metrics. We use `lock` to synchronize and ensure that Collect will only read the data after the `reader` has finished updating `containersMetrics`.

Note that having internal goroutines in the `Collect` function does not interfere with the `reader` function since the gorotines in the `Collect` function only read the data. So they will never create any race conditions!

Reply:
em ...
`reader` update `containersMetrics` in each 3s since 0s.
`Collect ` consume `containersMetrics` in each 3s since 2.x s.

in an edge case, `Collect ` WIP, and `reader` updated data. then ... the result of `Collect` will be two parts, one for before `reader` updates at 3s and the other is `reader` updates after 3s?

is that acceptable?

Reply:
The result of `Collect` will be only one part, after the `reader`, never in between....

The `reader` only start when the `Collect` has finished, and vice-versa....
So, in practice, the `Collect` will report the results of the previous `reader` iteration

Since the Collect function is very fast (specially because we have now gorotines) I expect that it will not change the `reader` sampling interval too much...but maybe we should log how long the `reader` was waiting to start